posts/2024-03-29-ArgoCD-Configuration-Drift.adoc

```diff
@@ -0,0 +1,163 @@
+```
+
+For a more detailed tutorial on the Open Liberty Operator and how to make use of it, see our interactive guide: https://openliberty.io/guides/openliberty-operator-intro.html[Deploying a microservice to Kubernetes using Open Liberty Operator].
+
+[#setup_argocd]
+=== Setup ArgoCD for effective GitOps and Continuous Deployment
+
+ArgoCD follows a GitOps pattern of using Git repositories as the source of truth for defining the desired application state, which we will use to deploy and monitor our Open Liberty configuration repository.  ArgoCD will help us identify any drift that occurs between the current cluster state and the desired state in the configuration repository.  
+
+To install ArgoCD simply run these commands from the https://argo-cd.readthedocs.io/en/stable/getting_started/[Quickstart section] of the docs:
+
+
+First, we'll use the following commands to create a new namespace named `argocd` and install ArgoCD by applying manifests to that namespace.
+```
+kubectl create namespace argocd
+kubectl apply -n argocd -f https://raw.githubusercontent.com/argoproj/argo-cd/stable/manifests/install.yaml
+```
+
+Next we'll get the default Admin credentials to access the ArgoCD user interface.  Using the UI is a great way to get started, allowing us to create our application and identify configuration drift in a visual way.
+To get the default Admin password run the following command:
+```
+kubectl get secret argocd-initial-admin-secret -n argocd -o jsonpath='{.data.password}' | base64 -d
+```
+To access the ArgoCD UI, go to the route of the `argocd-server` and log with the `admin` user and password.
+[.img_border_light]
+image::/img/blog/configuration-drift-1.png[Argo Login,width=70%,align="center"]
+
+Once logged in, you should see the ArgoCD landing page similar to the screenshot below:
+[.img_border_light]
+image::/img/blog/configuration-drift-2.png[Argo Home,width=70%,align="center"]
+
+[#setup_gitops_repo]
+== Setting up a GitOps Repository for your Application
+Now that you've successfully installed ArgoCD and gained access to its console, the next step is to set up a GitHub repository for ArgoCD to synchronize with.  In our example we have created a deployment directory in the application code repo which we will point ArgoCD to for deployment.  In our experience we found it better to separate the deployment configuration into a separate repository from the application code, however for this small example we have it with the application code.  
+
+It's a recommended best practice to maintain separation between configuration repositories and code repositories. Keeping these separate ensures that infrastructure changes are decoupled from application changes, allowing for more granular control, better audit trails, and reduced risk of unintended side-effects.  The reasons why can be found in https://argo-cd.readthedocs.io/en/stable/user-guide/best_practices/[ArgoCD's documentation on best practices].
```

Comment:
```suggestion
== Setting up a GitOps Repository for your Application
Now that you've successfully installed ArgoCD and gained access to its console, the next step is to set up a GitHub repository for ArgoCD to synchronize with.  In our example we have created a deployment directory in the application code repo which we will point ArgoCD to for deployment.  

_NOTE: In our experience we found it better to separate the deployment configuration into a separate repository from the application code. This aligns with recommended best practices to maintain separation between configuration repositories and code repositories. Keeping these separate ensures that infrastructure changes are decoupled from application changes, allowing for more granular control, better audit trails, and reduced risk of unintended side-effects.  The reasons why can be found in https://argo-cd.readthedocs.io/en/stable/user-guide/best_practices/[ArgoCD's documentation on best practices]. However, for this particular example, because it is just a small demo application, we have put the deployment configuration with the application code._
```